src/cx_ed25519.c

```diff
@@ -180,3 +187,168 @@ int scalarmult_ed25519(BIGNUM *Qx, BIGNUM *Qy, BIGNUM *Px, BIGNUM *Py, BIGNUM *e
 
   return scalarmult(&Q, &P, e);
 }
+
+void cx_encode_int(uint8_t *v, size_t len)
+{
+  uint8_t t;
+  int i, j;
+
+  j = len - 1;
+  len /= 2;
+
+  for (i = 0; len > 0; i++, j--, len--) {
+    t = v[i];
+    v[i] = v[j];
+    v[j] = t;
+    i++;
+    j--;
+  }
+}
+
+void cx_decode_int(uint8_t *v, size_t len)
+{
+  return cx_encode_int(v, len);
+}
+
+static void cx_compress(uint8_t *x, uint8_t *y, size_t size)
+{
+  if (x[size - 1] & 1) {
+    y[0] |= 0x80;
+  }
+
+  cx_encode_int(y, size);
+}
+
+int sys_cx_edward_compress_point(cx_curve_t curve, uint8_t *P, size_t P_len) {
+  cx_curve_twisted_edward_t *domain;
+  uint8_t *x, *y;
+  size_t size;
+
+  domain = (cx_curve_twisted_edward_t *)cx_ecfp_get_domain(curve);
+  size = domain->length;
+  if (curve != CX_CURVE_Ed25519 || P_len < (1 + 2 * size)) {
+    errx(1, "cx_edward_compress_point: invalid parameters");
+    return -1;
+  }
+
+  x = P + 1;
+  y = P + 1 + size;
+
+  cx_compress(x, y, size);
+  memcpy(x, y, size);
+
+  P[0] = 0x02;
+
+  return 0;
+}
+
+static int xrecover(BIGNUM *x, BIGNUM *y)
+{
+  BIGNUM *a, *xx;
+  int ret;
+
+  if (!initialized && initialize() != 0) {
+    return -1;
+  }
+
+  ret = 0;
+
+  a = BN_new();
+  xx = BN_new();
+  if (a == NULL || xx == NULL) {
+    ret = -1;
+    goto free_bn;
+  }
+
+  BN_mul(a, d1, y, ctx);
+  BN_mul(a, a, y, ctx);
+  BN_add(a, a, one);
+  BN_mod_inverse(a, a, q, ctx);
+  BN_mul(xx, y, y, ctx);
```

Comment:
This could be optimized, as ``y*y`` is computed twice (once in ``a = d1 * y * y``, and then here with ``xx = y * y``)